File: proyecto_en_java/src/On_logn_Quicksort.java
package proyectoEda;
import java.util.List;
import java.util.Random;

/**
 * =======================================================================
 * CLASE: On_logn_Quicksort 
 * =======================================================================
 * Implementación del Quicksort de Tres Vías, optimizado para manejar
 * grandes volúmenes de valores duplicados (utilidadNormalizada) sin
 * causar un StackOverflowError.
 */
public class On_logn_Quicksort { 

    private static final Random GENERADOR_ALEATORIO = new Random();

    /**
     * Método principal llamado por el Main. Inicia el ordenamiento.
     * @param lista La lista de objetos NormalizedUtility a ordenar.
     */
    public static void sort(List<NormalizedUtility> lista) { 
        ordenar(lista, 0, lista.size() - 1);
    }

    /**
     * El método recursivo principal del Quicksort de Tres Vías.
     * Divide la lista en tres zonas: Menores, Iguales y Mayores al pivote.
     */
    private static void ordenar(List<NormalizedUtility> lista, int indiceInicio, int indiceFin) {
        
        // Condición base: si ya no hay elementos o solo hay uno, se detiene.
        if (indiceInicio >= indiceFin) return; 

        // 1. Pivote Aleatorio: Se elige un pivote al azar para estabilidad.
        int indiceAleatorio = GENERADOR_ALEATORIO.nextInt(indiceFin - indiceInicio + 1) + indiceInicio;
        // Se coloca el pivote en el índice de inicio
        intercambiar(lista, indiceAleatorio, indiceInicio); 

        // El pivote se selecciona del índice 'indiceInicio'
        double valorPivote = lista.get(indiceInicio).getUtilidadNormalizada();
        
        // 2. TRES PUNTEROS:
        int indiceMenor = indiceInicio;    // lt (Límite superior de la zona <)
        int indiceMayor = indiceFin;       // gt (Límite inferior de la zona >)
        int i = indiceInicio + 1;          // Puntero de escaneo

        // 3. Partición de Tres Vías
        while (i <= indiceMayor) {
            double valorActual = lista.get(i).getUtilidadNormalizada();
            
            if (valorActual < valorPivote) {
                // El valor es menor: Se mueve a la zona de "menores" (lt)
                intercambiar(lista, indiceMenor++, i++);
            } else if (valorActual > valorPivote) {
                // El valor es mayor: Se mueve a la zona de "mayores" (gt)
                intercambiar(lista, i, indiceMayor--); 
            } else {
                // El valor es igual: Se queda en la zona central. Solo avanza el escáner.
                i++;
            }
        }
        
        // 4. Llamada Recursiva
        // Se llama recursivamente SOLO a las zonas < y > (excluyendo la zona =)
        ordenar(lista, indiceInicio, indiceMenor - 1);  // Zona < valorPivote
        ordenar(lista, indiceMayor + 1, indiceFin);     // Zona > valorPivote
    }

    /**
     * Método de utilidad para intercambiar dos elementos en la lista (in-place).
     */
    private static void intercambiar(List<NormalizedUtility> lista, int i, int j) {
        NormalizedUtility temporal = lista.get(i);
        lista.set(i, lista.get(j));
        lista.set(j, temporal);
    }
}

File: algorithms/onlogn_quicksort.py
import random
from typing import List, Any


# =======================================================================
# onlogn_quicksort
# =======================================================================
"""
Implementación del Quicksort de Tres Vías (3-way partition) de Dijkstra.

Optimizado para manejar grandes volúmenes de valores duplicados 
(utilidadNormalizada) 
"""
# =======================================================================

def intercambiar(lista: List[Any], i: int, j: int):
    """
    Método de utilidad para intercambiar dos elementos en la lista (in-place).
    
    Args:
        lista (List): La lista donde se realiza el intercambio.
        i (int): Índice del primer elemento.
        j (int): Índice del segundo elemento.
    """
    lista[i], lista[j] = lista[j], lista[i]


def ordenar_recursivo(lista: List[Any], indice_inicio: int, indice_fin: int):
    """
    El metodo recursivo principal del Quicksort de Tres Vias.

    Divide la lista en tres zonas: Menores (<), Iguales (=) y Mayores (>) al pivote.
    
    Args:
        lista (List): La porción de la lista a ordenar.
        indice_inicio (int): Límite inferior de la porción actual.
        indice_fin (int): Límite superior de la porción actual.
    """
    # 1. Condicion base: si ya no hay elementos o solo hay uno, se detiene.
    if indice_inicio >= indice_fin: 
        return

    # 2. Pivote Aleatorio: Se elige un pivote al azar para estabilidad 
    # y evitar el peor caso (O(N^2)) en listas pre-ordenadas.
    indice_aleatorio = random.randint(indice_inicio, indice_fin)
    
    # Se coloca el pivote en el indice de inicio para la particion
    intercambiar(lista, indice_aleatorio, indice_inicio)

    # El pivote se selecciona del indice 'indice_inicio'
    # Asumimos que los objetos tienen el método para obtener el valor de comparación.
    valor_pivote = lista[indice_inicio].get_utilidad_normalizada()
    
    # 3. TRES PUNTEROS 
    indice_menor = indice_inicio    
    indice_mayor = indice_fin       
    i = indice_inicio + 1           

    # 4. tres caminos
    while i <= indice_mayor:
        valor_actual = lista[i].get_utilidad_normalizada()
        
        if valor_actual < valor_pivote:
            # Caso 1: Menor que el pivote. 
            # Se intercambia con el primer elemento de la zona central (=) y se expande la zona <.
            intercambiar(lista, indice_menor, i)
            indice_menor += 1
            i += 1 
            
        elif valor_actual > valor_pivote:
            # Caso 2: Mayor que el pivote. 
            # Se intercambia con el último elemento de la zona > y se contrae la zona >.
            intercambiar(lista, i, indice_mayor)
            indice_mayor -= 1  
            # i no avanza porque el elemento recién intercambiado en lista[i] necesita ser revisado
            
        else:
            # Caso 3: Igual al pivote. 
            # Se queda en la zona central
            i += 1
    
    # 5. Llamada Recursiva
    # Se llama recursivamente SOLO a las zonas < y > (excluyendo la zona =)
    ordenar_recursivo(lista, indice_inicio, indice_menor - 1)  # Zona < valorPivote
    ordenar_recursivo(lista, indice_mayor + 1, indice_fin)      # Zona > valorPivote


def sort(lista: List[Any]):
    """
    Método principal (interfaz pública). Inicia el ordenamiento.
    
    Llama al método recursivo con los índices de toda la lista.
    
    Args:
        lista (List): La lista de objetos NormalizedUtility a ordenar (in-place).
    """
    ordenar_recursivo(lista, 0, len(lista) - 1)

